File: src/main/java/nl/hanze/StockApp.java
package nl.hanze;


import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.layout.FlowPane;
import javafx.scene.layout.Pane;
import javafx.scene.layout.StackPane;
import javafx.stage.Stage;
import nl.hanze.observers.AAPLObserver;
import nl.hanze.observers.GoogleObserver;
import nl.hanze.observers.IBMObserver;
import nl.hanze.view.GraphView;
import nl.hanze.view.TextView;

import java.util.Arrays;
import java.util.stream.Stream;


// the Client
public class StockApp extends Application {

    private View graphView;
    private View textView;
    private Pane root;
    private Pane viewPane;
    private Button textButton;
    private Button graphButton;

    public static void main(String[] args) {
        launch(args);
    }

    private static void spawnPriceGenerationThreads(StockGrabber stockGrabber) {
        Arrays.stream(Stock.values())
                .map(name -> new GetTheStock(stockGrabber, name))
                .map(Thread::new)
                .forEach(Thread::start);
    }

    @Override
    public void start(Stage primaryStage) throws Exception {
        StockGrabber stockGrabber = new StockGrabber();
        spawnPriceGenerationThreads(stockGrabber);
        graphView = new GraphView();
        textView = new TextView();
        createObservers(stockGrabber);
        root = new FlowPane();
        viewPane = new StackPane();
        createButtons();
        showView(primaryStage);
    }

    private void showView(Stage primaryStage) {
        root.getChildren().add(textButton);
        root.getChildren().add(graphButton);
        root.getChildren().add(viewPane);
        primaryStage.setScene(new Scene(root, 300, 250));
        primaryStage.show();
    }

    private void createObservers(StockGrabber stockGrabber) {
        AAPLObserver aaplObserver = new AAPLObserver(stockGrabber);
        GoogleObserver googleObserver = new GoogleObserver(stockGrabber);
        IBMObserver ibmObserver = new IBMObserver(stockGrabber);
        Arrays.asList(aaplObserver,googleObserver,ibmObserver)
                .forEach(observer ->{
                    observer.addView(textView);
                    observer.addView(graphView);
                });
    }

    private void createButtons() {
        textButton = new Button("Text");
        graphButton = new Button("Graph");
        textButton.setOnMouseClicked(t -> {
            viewPane.getChildren().clear();
            viewPane.getChildren().add(textView.getNode());
        });
        graphButton.setOnMouseClicked(t ->{
            viewPane.getChildren().clear();
            viewPane.getChildren().add(graphView.getNode());
        });
    }
}

File: src/main/java/nl/hanze/assignment1/StockApp.kt
package nl.hanze.assignment1


import javafx.application.Application
import javafx.event.EventHandler
import javafx.scene.Node
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.control.Button
import javafx.scene.input.MouseEvent
import javafx.scene.layout.FlowPane
import javafx.scene.layout.Pane
import javafx.scene.layout.StackPane
import javafx.stage.Stage
import nl.hanze.assignment1.observers.AAPLObserver
import nl.hanze.assignment1.observers.BaseObserver
import nl.hanze.assignment1.observers.GoogleObserver
import nl.hanze.assignment1.observers.IBMObserver
import nl.hanze.assignment1.view.GraphView
import nl.hanze.assignment1.view.TextView
import java.util.*


// the Client
class StockApp : Application() {

    private lateinit var graphView: View
    private lateinit var textView: View
    private lateinit var viewPane: Pane

    private fun Button.onClick(eventHandler: (MouseEvent) -> Unit) = apply {
        onMouseClicked = EventHandler<MouseEvent>(eventHandler)
    }

    private fun Stage.showScene(width: Number = -1, height: Number = -1, generator: () -> Parent) {
        scene = Scene(generator(), width.toDouble(), height.toDouble())
        show()
    }

    private fun flowPane(init: FlowPane.() -> Unit = {}) = FlowPane().also(init)
    private fun FlowPane.button(name: String, init: Button.() -> Unit = {}) = Button(name).also(init).also { children += it }
    inline fun <reified R : Node> FlowPane.add() = R::class.java.newInstance().also { children += it }

    @Throws(Exception::class)
    override fun start(primaryStage: Stage) {
        val stockGrabber = StockGrabber()
        spawnPriceGenerationThreads(stockGrabber)
        graphView = GraphView()
        textView = TextView()
        createObservers(stockGrabber)

        primaryStage.showScene(300, 256) {
            flowPane {
                button("text").onClick {
                    viewPane.children.setAll(textView.node)
                }
                button("Graph").onClick {
                    viewPane.children.setAll(graphView.node)
                }
                viewPane = add<StackPane>()
            }
        }
    }

    private fun createObservers(stockGrabber: StockGrabber) {
        val aaplObserver = AAPLObserver(stockGrabber)
        val googleObserver = GoogleObserver(stockGrabber)
        val ibmObserver = IBMObserver(stockGrabber)
        Arrays.asList<BaseObserver>(aaplObserver, googleObserver, ibmObserver)
                .forEach { observer ->
                    observer.addView(textView)
                    observer.addView(graphView)
                }
    }


    companion object {
        private fun spawnPriceGenerationThreads(stockGrabber: StockGrabber) = Stock.values().asSequence()
                .map { GetTheStock(stockGrabber, it) }
                .map(::Thread)
                .forEach { it.start() }
    }
}

fun main(args: Array<String>) = Application.launch(StockApp::class.java, *args)